src/services/heatmap.service.ts

```diff
@@ -0,0 +1,167 @@
+import { Connection } from 'mongoose';
+
+/**
+ * Create heatmap
+ * @param {IHeatMap} data
+ * @returns {Promise<IHeatMap>}
+ */
+async function getHeatmaps(connection: Connection, startDate: Date) {
+    const heatmaps = await connection.models.HeatMap.aggregate([
+        // Stage1 : find heatmaps between startDate and today
+        {
+            $match: {
+                'date': {
+                    $gte: new Date(startDate),
+                    $lte: new Date()
+                }
+            }
+        },
+
+        // Stage2 : extract needed data
+        {
+            $project: {
+                _id: 0,
+                'dayOfWeek': { $dayOfWeek: "$date" },
+                'hour': { $hour: "$date" },
+                'interactions': { '$sum': '$interactions' },
+            }
+        }
```

Comment:
The correct way to do it is something like this
```suggestion
       
  {
    '$match': {
      'date': {
        '$gte': new Date(startDate), 
        '$lte': new Date()
      }
    }
  }, {
    '$project': {
      'dayOfWeek': {
        '$dayOfWeek': '$date'
      }, 
      'messages': 1, 
      'interactions': 1, 
      'emojis': 1
    }
  }, {
    '$unwind': {
      'path': '$messages', 
      'includeArrayIndex': 'messages_index'
    }
  }, {
    '$unwind': {
      'path': '$interactions', 
      'includeArrayIndex': 'interactions_index'
    }
  }, {
    '$project': {
      'emojis': 1, 
      'messages': 1, 
      'interactions': 1, 
      'dayOfWeek': 1, 
      'hour': '$messages_index', 
      'compare': {
        '$cmp': [
          '$messages_index', '$interactions_index'
        ]
      }
    }
  }, {
    '$match': {
      'compare': 0
    }
  }, {
    '$unwind': {
      'path': '$emojis', 
      'includeArrayIndex': 'emojis_index'
    }
  }, {
    '$project': {
      'emojis': 1, 
      'messages': 1, 
      'interactions': 1, 
      'hour': 1, 
      'dayOfWeek': 1, 
      'compare': {
        '$cmp': [
          '$hour', '$emojis_index'
        ]
      }
    }
  }, {
    '$match': {
      'compare': 0
    }
  }, {
    '$group': {
      '_id': {
        'dayOfWeek': '$dayOfWeek', 
        'hour': '$hour'
      }, 
      'dayOfWeek': {
        '$first': '$dayOfWeek'
      }, 
      'hour': {
        '$first': '$hour'
      }, 
      'emojis': {
        '$sum': '$emojis'
      }, 
      'messages': {
        '$sum': '$messages'
      }, 
      'interactions': {
        '$sum': '$interactions'
      }
    }
  }

```

but It's too complex and too many iterations, hence I suggest you fetch the information from database using a match query and the iterate the documents based on their element.